pkg/replicator/scheme.go

```diff
@@ -0,0 +1,387 @@
+		rs.metrics.originIterations.Inc()
+
+		id, ok := thanosblock.IsBlockDir(name)
+		if !ok {
+			return nil
+		}
+
+		rs.metrics.originMetaLoads.Inc()
+		meta, metaNonExistentOrPartial, err := loadMeta(ctx, rs.fromBkt, id)
+		if metaNonExistentOrPartial {
+			// meta.json is the last file uploaded by a Thanos shipper,
+			// therefore a block may be partially present, but no meta.json
+			// file yet. If this is the case we skip that block for now.
+			rs.metrics.originPartialMeta.Inc()
+			level.Info(rs.logger).Log("msg", "block meta not uploaded yet. Skipping.", "block_uuid", id.String())
+			return nil
+		}
+		if err != nil {
+			return fmt.Errorf("load meta for block %v from origin bucket: %w", id.String(), err)
+		}
+
+		if len(meta.Thanos.Labels) == 0 {
+			// TODO(bwplotka): Allow injecting custom labels as shipper does.
+			level.Info(rs.logger).Log("msg", "block meta without Thanos external labels set. This is not allowed. Skipping.", "block_uuid", id.String())
+			return nil
+		}
+
+		level.Debug(rs.logger).Log("msg", "adding block to available blocks", "block_uuid", id.String())
+
+		availableBlocks = append(availableBlocks, meta)
+
+		return nil
+	}); err != nil {
+		return fmt.Errorf("iterate over origin bucket: %w", err)
+	}
+
+	candidateBlocks := []*metadata.Meta{}
+
+	for _, b := range availableBlocks {
+		if rs.blockFilter(b) {
+			level.Debug(rs.logger).Log("msg", "adding block to candidate blocks", "block_uuid", b.BlockMeta.ULID.String())
+			candidateBlocks = append(candidateBlocks, b)
+		}
+	}
+
+	// In order to prevent races in compactions by the target environment, we
+	// need to replicate oldest start timestamp first.
+	sort.Slice(candidateBlocks, func(i, j int) bool {
+		return candidateBlocks[i].BlockMeta.MinTime < candidateBlocks[j].BlockMeta.MinTime
+	})
+
+	for _, b := range candidateBlocks {
+		if err := rs.ensureBlockIsReplicated(ctx, b.BlockMeta.ULID); err != nil {
+			return fmt.Errorf("ensure block %v is replicated: %w", b.BlockMeta.ULID.String(), err)
+		}
+	}
+
+	return nil
+}
+
+// ensureBlockIsReplicated ensures that a block present in the origin bucket is
+// present in the target bucket.
+func (rs *replicationScheme) ensureBlockIsReplicated(ctx context.Context, id ulid.ULID) error {
+	blockID := id.String()
+	chunksDir := path.Join(blockID, thanosblock.ChunksDirname)
+	indexFile := path.Join(blockID, thanosblock.IndexFilename)
+	metaFile := path.Join(blockID, thanosblock.MetaFilename)
+
+	level.Debug(rs.logger).Log("msg", "ensuring block is replicated", "block_uuid", blockID)
+
+	originMetaFile, err := rs.fromBkt.Get(ctx, metaFile)
+	if err != nil {
+		return fmt.Errorf("get meta file from origin bucket: %w", err)
+	}
+
+	defer runutil.CloseWithLogOnErr(rs.logger, originMetaFile, "close original meta file")
+
+	targetMetaFile, err := rs.toBkt.Get(ctx, metaFile)
+	if targetMetaFile != nil {
+		defer runutil.CloseWithLogOnErr(rs.logger, targetMetaFile, "close target meta file")
```

Comment:
I guess now that we know that we have that invariant, we should continue the typical pattern here instead of this:
```
	targetMetaFile, err := rs.toBkt.Get(ctx, metaFile)
	if err != nil {
		return fmt.Errorf("get meta file from target bucket: %w", err)
	}
        defer runutil.CloseWithLogOnErr(rs.logger, targetMetaFile, "close target meta file")
        ...
```
Or is there some reason you wrote this in such a way?

Reply:
This `if` ensure targetMetaFile is not nil, the following `if` would handle error.